### DISCUSSION TITLE
Cross compilation for android ndk

### DISCUSSION THREAD

**Discussion by remul19** [fraunhoferhhi/vvdec#146] [Q&A] [answered]
Hi All,

The decoder code is working good on my Linux machine.
I am trying to cross compile this for android-ndk.
Could anyone please guide me on the steps for this cross-compiling?

**Comment by K-os** [comment, marked as answer]
Cross compilation for Android is straight forward using CMake as documented here:
https://developer.android.com/ndk/guides/cmake#command-line

For example to build for arm64/aarch64:
```
cmake -S . -B build/android -DCMAKE_TOOLCHAIN_FILE=${ANDROID_HOME}/ndk/25.2.9519653/build/cmake/android.toolchain.cmake -DANDROID_ABI=arm64-v8a  -DVVDEC_TOPLEVEL_OUTPUT_DIRS=0 -DCMAKE_BUILD_TYPE=Release
cmake --build build/android
```

The only thing specific to vvdec, that I am doing here is setting `VVDEC_TOPLEVEL_OUTPUT_DIRS=0`. That way the output will be written to `build/android/{bin,lib}/` instead of the `bin/` and `lib/` directories in the top level source dir, so it doesn't interfere with any builds for the current platform.

If you want to build a shared library, also specify `-DBUILD_SHARED_LIBS=1`.

**Reply by remul19** [reply]
Hi,

I downloaded android ndk Linux version from https://developer.android.com/ndk/downloads

The command used for building vvdec is `cmake -S . -B build/android -DCMAKE_TOOLCHAIN_FILE=/home/myTest/android_ndk/android-ndk-r25c/build/cmake/android.toolchain.cmake  -DANDROID_ABI=arm64-v8a  -DBUILD_SHARED_LIBS=1 -DVVDEC_TOPLEVEL_OUTPUT_DIRS=0 -DCMAKE_BUILD_TYPE=Release`

I am getting errors while building for android. Can you please guide me on what needs to be done? 
I am attaching the error logs for reference.  
[Error_logs_for_android_build.txt](https://github.com/fraunhoferhhi/vvdec/files/11936853/Error_logs_for_android_build.txt)

**Comment by K-os** [comment]
Oh, I just realized, that armv7 builds of the decoder crash on Android. So it only run in 64 bit mode on devices that support armv8 for now. I'll look into that, but I will probably not have the time before August or September.

**Comment by K-os** [comment]
> The command used for building vvdec is `cmake -S . -B build/android -DCMAKE_TOOLCHAIN_FILE=/home/myTest/android_ndk/android-ndk-r25c/build/cmake/android.toolchain.cmake -DANDROID_ABI=arm64-v8a -DBUILD_SHARED_LIBS=1 -DVVDEC_TOPLEVEL_OUTPUT_DIRS=0 -DCMAKE_BUILD_TYPE=Release`

I have no problem building with that NDK and CMake 3.26.4.

> I am getting errors while building for android. Can you please guide me on what needs to be done?

Your build log guides you on what needs to be done: please try using a recent CMake version.

**Reply by remul19** [reply]
Hi .. Thanks for you suggestions.
I tested on another Linux machine and it gave much better results. I used below commands.
```
cmake -S . -B build/android -DCMAKE_TOOLCHAIN_FILE=/home/myTest/android_ndk/android-ndk-r25c/build/cmake/android.toolchain.cmake -DANDROID_ABI=arm64-v8a -DVVDEC_TOPLEVEL_OUTPUT_DIRS=0 -DCMAKE_BUILD_TYPE=Release
cmake --build build/android
```

Build was going smooth, but at the very last stage, I got the below error.
```
/usr/bin/ld.gold: --no-rosegment: unknown option
/usr/bin/ld.gold: use the --help option for usage information
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

A screenshot of error message is attached for reference.
![image](https://github.com/fraunhoferhhi/vvdec/assets/136456486/52edd097-7cdb-4f93-b64c-1f68418313ca)

In some forums, people have mentioned about using` ld.gold ` but I could not get any valid solution for this issue. Can anyone please help me, to resolve this issue.

**Reply by remul19** [reply]
Hi ..  I upgraded to cmake v3.26.4  .. And now build is completed successfully.
Thanks for the support.

![image](https://github.com/fraunhoferhhi/vvdec/assets/136456486/077dcefb-1802-494d-b6a0-5b9572e63ae0)
